app/models/piece.rb

```diff
@@ -1,4 +1,84 @@
 class Piece < ApplicationRecord
   belongs_to :user
   belongs_to :game
+
+  # find the low end of the range of horizontal values to check; self.position + 1 excludes the current square
+  def x_begin(x)
+    [x, position_x + 1].min
+  end
+
+  # find the upper end of the range of horizontal values to check; position_x - 1 excludes the current square
+  def x_end(x)
+    [x, position_x - 1].max
+  end
+
+  # find the low end of the range of vertical values to check; position + 1 excludes the current square
+  def y_begin(y)
+    [y, position_y + 1].min
+  end
+
+  # find the upper end of the range of horizontal values to check; position_y - 1 excludes the current square
+  def y_end(y)
+    [y, position_y - 1].max
+  end
+
+  # set range helper methods
+  def x_range(x)
+    x_begin(x)..x_end(x)
+  end
+
+  def y_range(y)
+    y_begin(y)..y_end(y)
+  end
+
+  # check if a horizontal move is blocked
+  def check_horizontal_obstructions(x, y)
+    x_range(x).each do |x|
+      if game.piece_in_space?(x, y)
+        return true
+      end
+    end
+    return false
+  end
+
+  # check if a vertical move is blocked
+  def check_vertical_obstructions(x, y)
+    y_range(y).each do |y|
+      if game.piece_in_space?(x, y)
+        return true
+      end
+    end
+    return false
+  end
+
+  # check if a diagnol move is blocked
+  def check_diagonal_obstructions(x, y)
+    # for each value x in the x_range *combined with* with each value y from the y_range, check
+    # if there is a piece in the corresponding space
+    x_range(x).each do |x|
+      y_range(y).each do |y|
+        if game.piece_in_space?(x, y)
+          return true
+        end
+      end
+    end
+    return false
+  end
+
```

Comment:
Would this not check every square in a 2 dimensional grid in `x_range` and `y_range`? We should only check the squares in a diagonal line. This could be done in one loop instead a loop embedded within a loop. If the move is diagonal, then `x_range` = `y_range`. Therefore you can use one loop where `y_range = x_range`.

Reply:
I agree that checking only in a diagonal line is optimal; however, x_range and y_range won't always be equal. That will only be the case if the piece we want to move has `position_x` = `position_y` to begin with. I worked through this with a couple of examples, which you can see a visual representation of here: http://bit.ly/2Ntk9fg I am very open to thinking through the logic here to come up with something that checks only the desired diagonal line, though.

Reply:
I see what you are saying. I would be content submitting this as a base to work on but keeping in mind that the diagonal movement is incomplete. If for instance you want to move a bishop from A1 to H8, the `is_obstructed?` method will always return true because it's checking every square on the board. 

I am impressed with the cleanliness and simplicity of your code! This wasn't an easy thing to implement.

Reply:
Thanks! And great call on that move which will always return true. I'll be thinking about this and will submit another pull request with some refined code, hopefully today.